### PR TITLE
Fix mongo scram

### PR DESCRIPTION
**Pull request by re-Tick**
## Related Issue
  - Test mode fails for big java app that depends on documentDB and other reps.

Closes: #1301

#### Describe the changes you've made
Updates the extraction of client nonce from the saslStart request. And updates the conversationId in the response of saslStart and request of saslContinue.

For a client nonce in r like this: `n,,n=svoc-owner,r=9'sji{=${C"StM|SZWM1>Vpk`, the current extractClientNonce fails due to the presence of `=` char in the nonce. This is fixed by using `strings.TrimPrefix(part, "r=")` in place of the `strings.Split(part, "=")[1]` of the **extractClientNonce**. 

And the `saslContinue` is failing because we are returning the same conversationId in the response of `saslStart` due to which the key collision occurs for `authMessageMap` since, the key is used **conversationId** to get the authMessage for generating the server signature for response of saslContinue. So, the conversationId is updated for the response of **saslStart** and conversationId is extracted from the **saslContinue** request to fetch the authMessage from the  authMessageMap. 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation 

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

### CODE REVIEW

File: pkg/proxy/integrations/mongoparser/mongoparser.go [hunk 1]
```diff
@@ -596,7 +601,9 @@ func isHeartBeat(opReq Operation, requestHeader models.MongoHeader, mongoRequest
 	case wiremessage.OpMsg:
 		_, ok := mongoRequest.(*models.MongoOpMessage)
 		if ok {
-			return (opReq.IsIsAdminDB() && (strings.Contains(opReq.String(), "hello") || (isScramAuthRequest(mongoRequest.(*models.MongoOpMessage).Sections, logger)))) || opReq.IsIsMaster()
+			return (opReq.IsIsAdminDB() && strings.Contains(opReq.String(), "hello")) ||
```
Comment (charankamarapu) [inline pkg/proxy/integrations/mongoparser/mongoparser.go:604]: Please correct the indentation

File: pkg/proxy/integrations/mongoparser/mongoparser.go [hunk 2]
```diff
@@ -55,8 +55,9 @@ func (m *MongoParser) ProcessOutgoing(requestBuffer []byte, clientConn, destConn
 		m.logger.Debug("the outgoing mongo in record mode")
 		encodeOutgoingMongo(requestBuffer, clientConn, destConn, m.hooks, m.logger, ctx)
 	case models.MODE_TEST:
+		logger := m.logger.With(zap.Any("Client IP Address", clientConn.RemoteAddr().String()), zap.Any("Client ConnectionID", util.GetNextID()), zap.Any("Destination ConnectionID", util.GetNextID()))
```
Comment (charankamarapu) [inline pkg/proxy/integrations/mongoparser/mongoparser.go:58]: Is this change required? Looks like you are adding few fields to the logger by default. So will these default fields appear in info logs too..?

File: pkg/proxy/integrations/mongoparser/scramAuth.go [hunk 3]
```diff
@@ -22,14 +23,19 @@ func isScramAuthRequest(actualRequestSections []string, logger *zap.Logger) bool
 			return false
 		}
 
+		conversationId, _ := extractConversationId(actualMsg)
 		// Check if the message is for starting the SASL (authentication) process
 		if _, exists := actualMsg["saslStart"]; exists {
-			logger.Info("the recieved request is saslStart", zap.Any("OpMsg", actualMsg))
+			logger.Info("the recieved request is saslStart",
```
Comment (charankamarapu) [inline pkg/proxy/integrations/mongoparser/scramAuth.go:29]: shouldn't this be a debug log..?

File: pkg/proxy/integrations/mongoparser/scramAuth.go [hunk 4]
```diff
@@ -22,14 +23,19 @@ func isScramAuthRequest(actualRequestSections []string, logger *zap.Logger) bool
...
+				zap.Any("OpMsg", actualMsg),
+				zap.Any("conversationId", conversationId))
 			return true
 			// Check if the message is for final request of the SASL (authentication) process
 		} else if _, exists := actualMsg["saslContinue"]; exists {
-			logger.Info("the recieved request is saslContinue", zap.Any("OpMsg", actualMsg))
-
+			logger.Info("the recieved request is saslContinue",
```
Comment (charankamarapu) [inline pkg/proxy/integrations/mongoparser/scramAuth.go:35]: shouldn't this be a debug log..?

File: pkg/proxy/integrations/scram/util.go [hunk 5]
```diff
@@ -29,7 +29,11 @@ func extractClientNonce(firstMsg string) (string, error) {
 	for _, part := range parts {
 		if strings.HasPrefix(part, "r=") {
 			// Split based on "=" and get the value of "r"
-			value := strings.Split(part, "=")[1]
+			// value := strings.Split(part, "=")[1]
+			value := strings.TrimPrefix(part, "r=")
+			if value == part {
```
Comment (charankamarapu) [inline pkg/proxy/integrations/scram/util.go:34]: Is there any chance that this case occurs..?

File: pkg/proxy/integrations/mongoparser/scramAuth.go [hunk 6]
```diff
@@ -266,6 +292,11 @@ func handleSaslStart(i int, actualMsg map[string]interface{}, expectedRequestSec
 	logger.Debug("after replacing the new client nonce in auth response", zap.String("first response", newFirstAuthResponse))
 	// replace the payload with new first response auth
 	responseMsg["payload"].(map[string]interface{})["$binary"].(map[string]interface{})["base64"] = base64.StdEncoding.EncodeToString([]byte(newFirstAuthResponse))
+	responseMsg, err = updateConversationId(responseMsg, int(util.GetNextID()))
```
Comment (charankamarapu) [inline pkg/proxy/integrations/mongoparser/scramAuth.go:295]: For my clarity, the thing we should consider here mainly is that, the conversationId shouldn't be same. Every time it needs to change. Is this the only change done here ? and doesn't that depend on the client request message ..? Some more info regarding conversationId helps me understand  more regarding this change mainly like why it is used etc.

**Review by charankamarapu — COMMENTED** [review]
Please address the comments. Most of the comments are questions for my understanding.

File: pkg/proxy/integrations/mongoparser/mongoparser.go (hunk 2 above)
Comment (re-Tick) [reply pkg/proxy/integrations/mongoparser/mongoparser.go:58]: Yes these fields are required and they will only appear for error and debug logs of decodeOutgoingMongo function only. Other info logs will not be effected by this

File: pkg/proxy/integrations/scram/util.go (hunk 5 above)
Comment (re-Tick) [reply pkg/proxy/integrations/scram/util.go:34]: Yes I have mentioned an example for this in the PR comments.

File: pkg/proxy/integrations/mongoparser/mongoparser.go (hunk 2 above)
Comment (charankamarapu) [reply pkg/proxy/integrations/mongoparser/mongoparser.go:58]: ok got it.

File: pkg/proxy/integrations/scram/util.go (hunk 5 above)
Comment (charankamarapu) [reply pkg/proxy/integrations/scram/util.go:34]: Yeah I have seen that but that is example for difference between `strings.TrimPrefix(part, "r=")` and `strings.Split(part, "=")[1]` right ? My comment was regarding even after using TrimPrefix will `value == part` anytime ?

File: pkg/proxy/integrations/mongoparser/scramAuth.go (hunk 6 above)
Comment (re-Tick) [reply pkg/proxy/integrations/mongoparser/scramAuth.go:295]: Since scram is a challenge based authentication, so the challenge is done in conversation format. There are 2 request-response cycles: saslStart and saslContinue. During the response of saslStart the conversationId is passed by db server, and that is used by the saslContinue request. Using this conversationId the authMessage are mapped to generate the server proof. So this has to be unique so that we can map concurrent authMessage's. Earlier the field is not updated for the saslStart response.

Comment (charankamarapu) [reply pkg/proxy/integrations/mongoparser/scramAuth.go:295]: Got it.

File: pkg/proxy/integrations/mongoparser/mongoparser.go (hunk 1 above)
Comment (re-Tick) [reply pkg/proxy/integrations/mongoparser/mongoparser.go:604]: Indentation is correct. I have done this because line was getting bigger. And it is a good practice.☺️

Comment (charankamarapu) [reply pkg/proxy/integrations/mongoparser/mongoparser.go:604]: Got it.

File: pkg/proxy/integrations/scram/util.go (hunk 5 above)
Comment (re-Tick) [reply pkg/proxy/integrations/scram/util.go:34]: That is a check, which can be true if the decoding of base64 of a wrong payload. Using this the error can be handled.

Comment (charankamarapu) [reply pkg/proxy/integrations/scram/util.go:34]: I mean there is a `if strings.HasPrefix(part, "r=")` then we used `strings.TrimPrefix(part, "r=")` how can value == part..?

**Review by gouravkrosx — APPROVED** [review]
LGTM

**Comment by charankamarapu** [comment]
I think there are few other changes added in the PR from other's commit.

**Comment by re-Tick** [comment]
> I think there are few other changes added in the PR from other's commit.

Yes they were made to merge to main branch. but there is some issue with the main branch.
